prebid-server/pbs-modules/rules-engine.md

```diff
@@ -0,0 +1,805 @@
+
+Mapping these concepts to the JSON syntax:
+
+- Each level of the tree corresponds to an entry in the 'schema' JSON element.
+- Each branch of the tree is defined by the 'conditions' element in the JSON and compared against the value returned by the [schema function](#schema-functions)
+- The contents of each leaf are stored in the 'results' JSON element, which is one or more [results functions](#results-functions)
+
+{: .alert.alert-warning :}
+This model differs from how rules are handled in the [floors system](/dev-docs/modules/floors.html#rule-handling). In that model, the runtime system has the ability to "go back up" the tree if it finds a dead-end. While friendlier, that model can create performance problems. The Rules Engine system is a little harder to utilize, but it guarantees that each dimension is processed only once.
+
+Here's an example rule tree:
+
+![Rules Engine Rules Tree](/assets/images/prebid-server/pbs_rules_engine_tree.png){:class="pb-xlg-img"}
+
+1. The first level is 'Dimension 1'. It has 3 specific values and one wildcard "other" value.
+2. The second level is 'Dimension 2'. It has 2 or 3 values for each scenario and possibly a wildcard value.
+3. There are 12 specific leaf nodes and one overall default leaf.
+4. Note that the second branch does not have a wildcard option at the second level, so if none of the 3 specific values for Dimension 2 match (red lines), the module will jump to the default leaf.
+
+{: .alert.alert-info :}
+Tip: you can also think of the rule tree as an N-dimensional matrix where each intersection of schema function values resolves to a 'results' node.
+
+See the [runtime processing](#runtime-processing) section below for more details.
+
+### Schema Functions
+
+The module provides a set of functions that parse the OpenRTB request for specific fields, extracting a particular value. For instance, finding out which specific country or channel is associated with this request. These functions are called 'schema functions' because they form the structure of the Rule Tree as described in the section above.
+
+All of these functions operate only on the OpenRTB request at the 'Processed Auction Request' stage of module invocation.
+
+{: .table .table-bordered .table-striped }
+| Schema Function | Args | Description | Output Example |
+| :---- | :---- | :---- | :---- |
+| deviceCountry | null  | Returns OpenRTB device.geo.country. | "CAN" |
+| deviceCountryIn | array of strings representing country codes | If the parameter is an array of strings, then check to see if device.geo.country is part of the named set (case sensitive). If yes, then return "true", else return "false". The parameter must be an array of strings. No length limit. If the array is length zero, treat it as 'null' above. | "true" |
```

Comment:
I think we can simplify this as I see an empty array or missing array arg as an error. Instead the user should just use `deviceCountry` with a condition of "*" or "" to achieve this behavior. I'm wondering if you want this function to handle this case because it makes it easier for a machine learning algorithm somehow by not having to possibly switch back and forth between `deviceCountry` and `deviceCountryIn` to address the empty case?

Reply:
Allowing "*" as a value in the array could be useful. @And1sS - what's implemented in PBS-J?